// @dicebear/converter@9.4.2/lib/core.js
import { getMimeType } from './utils/mime-type.js';
import { ensureSize } from './utils/svg.js';
export const toPng = (avatar, options = {}) => {
    return toFormat(avatar, 'png', options);
};
export const toJpeg = (avatar, options = {}) => {
    return toFormat(avatar, 'jpeg', options);
};
export const toWebp = (avatar, options = {}) => {
    return toFormat(avatar, 'webp', options);
};
export const toAvif = (avatar, options = {}) => {
    return toFormat(avatar, 'avif', options);
};
function toFormat(avatar, format, options = {}) {
    if (options.includeExif) {
        console.warn('The `exif` option is not supported in the browser version of `@dicebear/converter`. \n' +
            'Please use the node version of `@dicebear/converter` to generate images with exif data.');
    }
    const svg = typeof avatar === 'string' ? avatar : avatar.toString();
    return {
        toDataUri: () => toDataUri(svg, format, options),
        toArrayBuffer: () => toArrayBuffer(svg, format, options),
    };
}
async function toDataUri(svg, format, options) {
    if ('svg' === format) {
        return `data:${getMimeType(format)};utf8,${encodeURIComponent(svg)}`;
    }
    const canvas = await toCanvas(svg, format, options);
    return canvas.toDataURL(getMimeType(format));
}
async function toArrayBuffer(rawSvg, format, options) {
    const canvas = await toCanvas(rawSvg, format, options);
    return await new Promise((resolve, reject) => {
        canvas.toBlob((blob) => {
            blob
                ? resolve(blob.arrayBuffer())
                : reject(new Error('Could not create blob'));
        }, getMimeType(format));
    });
}
async function toCanvas(rawSvg, format, options) {
    const { svg, size } = ensureSize(rawSvg, options.size);
    const canvas = document.createElement('canvas');
    canvas.width = size;
    canvas.height = size;
    const context = canvas.getContext('2d');
    if (null === context) {
        throw new Error('Could not get canvas context');
    }
    if (format === 'jpeg') {
        context.fillStyle = 'white';
        context.fillRect(0, 0, size, size);
    }
    const img = document.createElement('img');
    img.width = size;
    img.height = size;
    img.setAttribute('src', await toDataUri(svg, 'svg', options));
    return new Promise((resolve, reject) => {
        img.onload = () => {
            context.drawImage(img, 0, 0, size, size);
            resolve(canvas);
        };
        img.onerror = (e) => reject(e);
    });
}
