// @dicebear/converter@9.4.2/lib/node/core.js
import { promises as fs } from 'node:fs';
import { getMimeType } from '../utils/mime-type.js';
import { ensureSize, getMetadata } from '../utils/svg.js';
import * as tmp from 'tmp-promise';
import { renderAsync } from '@resvg/resvg-js';
import sharp from 'sharp';
import { exiftool } from 'exiftool-vendored';
export const toPng = (avatar, options = {}) => {
    return toFormat(avatar, 'png', options);
};
export const toJpeg = (avatar, options = {}) => {
    return toFormat(avatar, 'jpeg', options);
};
export const toWebp = (avatar, options = {}) => {
    return toFormat(avatar, 'webp', options);
};
export const toAvif = (avatar, options = {}) => {
    return toFormat(avatar, 'avif', options);
};
function toFormat(avatar, format, options) {
    var _a;
    const svg = typeof avatar === 'string' ? avatar : avatar.toString();
    const exifOption = (_a = options.includeExif) !== null && _a !== void 0 ? _a : false;
    const exif = exifOption ? getExif(svg) : {};
    return {
        toDataUri: () => toDataUri(svg, format, exif, options),
        toArrayBuffer: () => toArrayBuffer(svg, format, exif, options),
    };
}
async function toDataUri(svg, format, exif, options) {
    if (format === 'svg') {
        return `data:${getMimeType(format)};utf8,${encodeURIComponent(svg)}`;
    }
    const buffer = await toBuffer(svg, format, exif, options);
    return `data:${getMimeType(format)};base64,${buffer.toString('base64')}`;
}
async function toArrayBuffer(rawSvg, format, exif, options) {
    return (await toBuffer(rawSvg, format, exif, options)).buffer;
}
async function toBuffer(rawSvg, format, exif, options) {
    const hasFonts = Array.isArray(options.fonts);
    const { svg, size } = ensureSize(rawSvg, options.size);
    let buffer = (await renderAsync(svg, {
        fitTo: { mode: 'width', value: size },
        font: {
            loadSystemFonts: !hasFonts,
            fontFiles: hasFonts ? options.fonts : undefined,
        },
    })).asPng();
    if (format !== 'png') {
        const sharpInstance = sharp(buffer);
        if (format === 'jpeg') {
            sharpInstance.flatten({ background: '#ffffff' });
        }
        buffer = await sharpInstance.toFormat(format).toBuffer();
    }
    if (Object.keys(exif).length > 0) {
        buffer = await tmp.withFile(async ({ path }) => {
            await fs.writeFile(path, buffer);
            await exiftool.write(path, exif, { writeArgs: ['-overwrite_original'] });
            return fs.readFile(path);
        });
    }
    return buffer;
}
// https://www.iptc.org/std/photometadata/specification/IPTC-PhotoMetadata
// https://developers.google.com/search/docs/appearance/structured-data/image-license-metadata
function getExif(svg) {
    const metadata = getMetadata(svg);
    const exif = {};
    if (metadata.title) {
        exif['IPTC:ObjectName'] = metadata.title;
        exif['XMP-dc:Title'] = metadata.title;
    }
    if (metadata.source) {
        exif['XMP-plus:LicensorURL'] = metadata.source;
    }
    if (metadata.creator) {
        exif['IPTC:By-line'] = metadata.creator;
        exif['XMP-dc:Creator'] = metadata.creator;
        exif['IPTC:Credit'] = metadata.creator;
        exif['XMP-photoshop:Credit'] = metadata.creator;
    }
    if (metadata.license) {
        exif['XMP-xmpRights:WebStatement'] = metadata.license;
    }
    if (metadata.copyright) {
        exif['IPTC:CopyrightNotice'] = metadata.copyright;
        exif['XMP-dc:Rights'] = metadata.copyright;
    }
    return exif;
}
